apigee/registry
# Implement Spectral Linter Plugin

## Summary
This pull request implements the Spectral Linter plugin to lint OpenAPI v2/v3 and AsyncAPI v2 API Specs. In addition, there are slight modifications that I have made to the API Linter interface for clarity.


On any given lint request, the linter will create the appropriate configuration based on the mime type of the spec that needs to be linted.

## Test
- Have registry running on `localhost:8080`.
- Ensure that https://github.com/APIs-guru/openapi-directory is extracted to your Desktop.
- Ensure that the following style guide is located under `~/styleguide.yaml`:
```yaml
name: openapistyleguide
mime_types:
  - application/x.openapi+gzip;version=2
  - application/x.openapi+gzip;version=3
guidelines:
  - name: refproperties
    display_name: Govern Ref Properties
    description: This guideline governs properties for ref fields on specs.
    rules:
      - name: norefsiblings
        description: An object exposing a $ref property cannot be further extended with additional properties.
        linter: spectral
        linter_rulename: no-$ref-siblings
        severity: 1
    status: 2
linters:
  - name: spectral
    uri: https://github.com/stoplightio/spectral

```

Now, run the following series of bash commands:
```bash
apg registry delete-project --name projects/example

apg registry create-project --project_id example --insecure --address localhost:8080

registry upload bulk openapi --project_id example ~/Desktop/openapi-directory/APIs

registry upload styleguide ~/styleguide.yaml --project_id=example

registry compute conformance projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml

registry get projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml/artifacts/lint-spectral --contents
```

The result should be a populated list of data which is the contents of the artifact attached to the spec. This data, when marshalled appropriately, contains a `LintResult` object attached to the spec.
